packages/commonwealth/server/util/cosmosProxy.ts

```diff
@@ -67,47 +110,150 @@ function setupCosmosProxy(
         });
         if (!chain) {
-          throw new AppError('Invalid chain');
+          throw new AppError(`Invalid chain: ${chainParam}`);
         }
-        const targetUrl = chain.ChainNode?.alt_wallet_url;
-        if (!targetUrl) {
-          throw new AppError('No LCD endpoint found');
+
+        let response;
+        const chainNode = chain.ChainNode;
+        cosmos_chain_id = chainNode.cosmos_chain_id;
+        chainNodeRestUrl = chainNode.alt_wallet_url?.trim();
+        let useProxy = !chainNodeRestUrl;
+        previouslyFailed = chainNode?.health === NodeHealth.Failed;
+
+        if (previouslyFailed) {
+          const lastUpdate = chainNode.updated_at?.getTime();
+          const healthPauseTimeout = new Date(
+            lastUpdate + FALLBACK_NODE_DURATION,
+          );
+          useProxy = previouslyFailed && new Date() > healthPauseTimeout;
         }
-        log.trace(`Found cosmos endpoint: ${targetUrl}`);
-        const rewrite = req.originalUrl.replace(req.baseUrl, targetUrl);
-        const body = _.isEmpty(req.body) ? null : req.body;
-
-        const response = await axios.post(rewrite, body, {
-          headers: {
-            origin: 'https://commonwealth.im',
-          },
-        });
-        log.trace(
-          `Got response from endpoint: ${JSON.stringify(
-            response.data,
-            null,
-            2,
-          )}`,
-        );
+
+        if (chainNodeRestUrl && !useProxy) {
+          log.trace(`Found cosmos endpoint: ${chainNodeRestUrl}`);
+          const rewrite = req.originalUrl.replace(
+            req.baseUrl,
+            chainNodeRestUrl,
+          );
+          const body = _.isEmpty(req.body) ? null : req.body;
+
+          response = await axios
+            .post(rewrite, body, {
+              headers: {
+                origin: 'https://commonwealth.im',
+              },
+            })
+            .catch(async (err) => {
+              log.trace(`Error: ${err.message}`);
+              await flagFailedNode(
+                previouslyFailed,
+                cosmos_chain_id,
+                chainNodeRestUrl,
+              );
+              return queryRESTProxy(req, cosmos_chain_id);
+            });
+
+          await flagHealthyNode(
+            response,
+            previouslyFailed,
+            cosmos_chain_id,
+            chainNodeRestUrl,
+          );
+        } else {
+          response = await queryRESTProxy(req, cosmos_chain_id);
+        }
+
+        log.trace(`Got response: ${JSON.stringify(response.data, null, 2)}`);
         return res.send(response.data);
       } catch (err) {
-        res.status(500).json({ message: err.message });
+        await flagFailedNode(
+          previouslyFailed,
+          cosmos_chain_id,
+          chainNodeRestUrl,
+        );
+        res.status(500).json({
+          message: err.message,
+        });
       }
     },
   );
 
+  const queryRESTProxy = async (req, cosmos_chain_id) => {
```

Comment:
Something strange is going on here. We have the REST proxy but we never use it? Also if only the rest/rpc part of the url changes, then maybe we should have `queryProxy(req, cosmos_chain_id, webProtocol: 'rpc' | 'rest')`.

Reply:
Good call. now `queryExternalProxy`